src/main/java/bio/terra/workspace/service/datareference/model/ReferenceObject.java

```diff
@@ -0,0 +1,53 @@
+package bio.terra.workspace.service.datareference.model;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeInfo.As;
+import com.fasterxml.jackson.annotation.JsonTypeInfo.Id;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * An interface representing the subject of a data reference.
+ *
+ * <p>All referenced objects are read-only value objects described by a set of string key-value
+ * pairs. The set of keys available are specific to each referenced object type, as are validation
+ * rules. ReferenceObject implementations must be fully serializable by the output of getProperties.
+ *
+ * <p>In order to properly deserialize implementations of this class, the implementation classes
+ * must be added to the list inside the {@Code JsonSubTypes} annotation below. The implementation
+ * class should specify a name using the {@Code JsonTypeName} annotation so that changes to the
+ * class name do not break backwards compatibility.
+ */
+@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY)
+@JsonSubTypes({@JsonSubTypes.Type(value = SnapshotReference.class, name = "SnapshotReference")})
+public interface ReferenceObject {
+
+  /**
+   * This objectMapper is static to ReferenceObjects so changes to other object mappers will not
+   * affect serialization of these objects.
+   */
+  ObjectMapper objectMapper = new ObjectMapper();
+
+  /**
+   * Method for serializing a ReferenceObject to a json string.
+   *
+   * <p>This method is used for Stairway and database serialization.
+   */
+  String toJson();
+
+  /**
+   * Method for deserializing a ReferenceObject from a json string.
+   *
+   * <p>This method is used for Stairway and database deserialization. This makes use of type
+   * information stored via the {@Code JsonTypeInfo} annotation, and parses that information using
+   * the {@Code JsonSubTypes} annotation.
+   */
+  static ReferenceObject fromJson(String jsonString) {
+    try {
+      return objectMapper.readValue(jsonString, ReferenceObject.class);
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException("Unable to deserialize ReferenceObject from string " + jsonString);
```

Comment:
I'm moderately worried about including the raw `jsonString` in the exception here. It could include sensitive information, or it could be large and unwieldy. How about logging the error and the raw string,  but only including a generic message in the thrown exception?

Reply:
Sounds good. I think this qualifies as an error, so I'm logging at that level